File: csharp/src/Vista.SDK/Experimental/PMSLocalIdBuilder.cs
using System.Text;
using Vista.SDK.Internal;

namespace Vista.SDK.Experimental;

public sealed partial record class PMSLocalIdBuilder : ILocalIdBuilder<PMSLocalIdBuilder, PMSLocalId>
{
    public static readonly string NamingRule = "dnv-v2-experimental";
    public static readonly CodebookName[] UsedCodebooks = new[]
    {
        CodebookName.Quantity,
        CodebookName.Content,
        CodebookName.State,
        CodebookName.Command,
        CodebookName.FunctionalServices,
        CodebookName.MaintenanceCategory,
        CodebookName.ActivityType,
        CodebookName.Position,
        CodebookName.Detail,
    };
    public VisVersion? VisVersion { get; private init; }

    public bool VerboseMode { get; private init; } = false;

    internal LocalIdItems Items { get; private init; }

    public GmodPath? PrimaryItem => Items.PrimaryItem;

    public GmodPath? SecondaryItem => Items.SecondaryItem;

    public MetadataTag? Quantity { get; private init; }

    public MetadataTag? Content { get; private init; }

    public MetadataTag? State { get; private init; }

    public MetadataTag? Command { get; private init; }

    public MetadataTag? FunctionalServices { get; private init; }

    public MetadataTag? MaintenanceCategory { get; private init; }

    public MetadataTag? ActivityType { get; private init; }

    public MetadataTag? Position { get; private init; }

    public MetadataTag? Detail { get; private init; }

    public PMSLocalIdBuilder WithVisVersion(in string visVersion)
    {
        var localIdbuilder = TryWithVisVersion(visVersion, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(visVersion));

        return localIdbuilder;
    }

    public PMSLocalIdBuilder WithVisVersion(in VisVersion version)
    {
        var localIdbuilder = TryWithVisVersion(version, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(WithVisVersion));

        return localIdbuilder;
    }

    public PMSLocalIdBuilder TryWithVisVersion(in VisVersion? visVersion) => TryWithVisVersion(visVersion, out _);

    public PMSLocalIdBuilder TryWithVisVersion(in string? visVersionStr, out bool succeeded)
    {
        if (VisVersions.TryParse(visVersionStr, out VisVersion v) == true)
        {
            var localIdbuilder = TryWithVisVersion(v, out var succeededInner);
            succeeded = succeededInner;
            return localIdbuilder;
        }
        succeeded = false;
        return this;
    }

    public PMSLocalIdBuilder TryWithVisVersion(in VisVersion? visVersion, out bool succeeded)
    {
        succeeded = true;
        return this with { VisVersion = visVersion };
    }

    public PMSLocalIdBuilder WithoutVisVersion() => this with { VisVersion = null };

    public PMSLocalIdBuilder WithVerboseMode(in bool verboseMode) => this with { VerboseMode = verboseMode };

    public PMSLocalIdBuilder WithPrimaryItem(in GmodPath item)
    {
        var localIdbuilder = TryWithPrimaryItem(item, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(WithPrimaryItem));

        return localIdbuilder;
    }

    public PMSLocalIdBuilder TryWithPrimaryItem(in GmodPath? item) => TryWithPrimaryItem(item, out _);

    public PMSLocalIdBuilder TryWithPrimaryItem(in GmodPath? item, out bool succeeded)
    {
        if (item is null)
        {
            succeeded = false;
            return this;
        }

        succeeded = true;
        return this with { Items = this.Items with { PrimaryItem = item } };
    }

    public PMSLocalIdBuilder WithoutPrimaryItem() => this with { Items = this.Items with { PrimaryItem = null } };

    public PMSLocalIdBuilder WithSecondaryItem(in GmodPath item)
    {
        var localIdBuilder = TryWithSecondaryItem(item, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(WithSecondaryItem));

        return localIdBuilder;
    }

    public PMSLocalIdBuilder TryWithSecondaryItem(in GmodPath? item) => TryWithSecondaryItem(in item, out _);

    public PMSLocalIdBuilder TryWithSecondaryItem(in GmodPath? item, out bool succeeded)
    {
        if (item is null)
        {
            succeeded = false;
            return this;
        }

        succeeded = true;
        return this with { Items = this.Items with { SecondaryItem = item } };
    }

    public PMSLocalIdBuilder WithoutSecondaryItem() => this with { Items = this.Items with { SecondaryItem = null } };

    public PMSLocalIdBuilder WithMetadataTag(in MetadataTag metadataTag)
    {
        var localIdBuilder = TryWithMetadataTag(metadataTag, out var succeeded);
        if (!succeeded)
            throw new ArgumentException("invalid metadata codebook name: " + metadataTag.Name);
        return localIdBuilder;
    }

    public PMSLocalIdBuilder TryWithMetadataTag(in MetadataTag? metadataTag) => TryWithMetadataTag(metadataTag, out _);

    public PMSLocalIdBuilder TryWithMetadataTag(in MetadataTag? metadataTag, out bool succeeded)
    {
        if (metadataTag is null)
        {
            succeeded = false;
            return this;
        }

        switch (metadataTag.Value.Name)
        {
            case CodebookName.Quantity:
                succeeded = true;
                return WithQuantity(metadataTag.Value);
            case CodebookName.Content:
                succeeded = true;
                return WithContent(metadataTag.Value);

            case CodebookName.State:
                succeeded = true;
                return WithState(metadataTag.Value);
            case CodebookName.Command:
                succeeded = true;
                return WithCommand(metadataTag.Value);
            case CodebookName.MaintenanceCategory:
                succeeded = true;
                return WithMaintenanceCategory(metadataTag.Value);
            case CodebookName.FunctionalServices:
                succeeded = true;
                return WithFunctionalServices(metadataTag.Value);
            case CodebookName.ActivityType:
                succeeded = true;
                return WithActivityType(metadataTag.Value);
            case CodebookName.Position:
                succeeded = true;
                return WithPosition(metadataTag.Value);
            case CodebookName.Detail:
                succeeded = true;
                return WithDetail(metadataTag.Value);
            default:
                succeeded = false;
                return this;
        }
    }

    public PMSLocalIdBuilder WithoutMetadataTag(in CodebookName name) =>
        name switch
        {
            CodebookName.Quantity => WithoutQuantity(),
            CodebookName.Content => WithoutContent(),
            CodebookName.State => WithoutState(),
            CodebookName.Command => WithoutCommand(),
            CodebookName.FunctionalServices => WithoutFunctionalServices(),
            CodebookName.MaintenanceCategory => WithoutMaintenanceCategory(),
            CodebookName.ActivityType => WithoutActivityType(),
            CodebookName.Position => WithoutPosition(),
            CodebookName.Detail => WithoutDetail(),
            _ => this
        };

    public PMSLocalIdBuilder WithoutQuantity() => this with { Quantity = null };

    public PMSLocalIdBuilder WithoutContent() => this with { Content = null };

    public PMSLocalIdBuilder WithoutState() => this with { State = null };

    public PMSLocalIdBuilder WithoutCommand() => this with { Command = null };

    public PMSLocalIdBuilder WithoutFunctionalServices() => this with { FunctionalServices = null };

    public PMSLocalIdBuilder WithoutMaintenanceCategory() => this with { MaintenanceCategory = null };

    public PMSLocalIdBuilder WithoutActivityType() => this with { ActivityType = null };

    public PMSLocalIdBuilder WithoutPosition() => this with { Position = null };

    public PMSLocalIdBuilder WithoutDetail() => this with { Detail = null };

    private PMSLocalIdBuilder WithQuantity(in MetadataTag quantity) => this with { Quantity = quantity, };

    private PMSLocalIdBuilder WithContent(in MetadataTag content) => this with { Content = content, };

    private PMSLocalIdBuilder WithState(in MetadataTag state) => this with { State = state, };

    private PMSLocalIdBuilder WithCommand(in MetadataTag command) => this with { Command = command, };

    private PMSLocalIdBuilder WithFunctionalServices(in MetadataTag functionalServices) =>
        this with
        {
            FunctionalServices = functionalServices,
        };

    private PMSLocalIdBuilder WithMaintenanceCategory(in MetadataTag maintenanceCategory) =>
        this with
        {
            MaintenanceCategory = maintenanceCategory,
        };

    private PMSLocalIdBuilder WithActivityType(in MetadataTag activityType) =>
        this with
        {
            ActivityType = activityType,
        };

    private PMSLocalIdBuilder WithPosition(in MetadataTag position) => this with { Position = position, };

    private PMSLocalIdBuilder WithDetail(in MetadataTag detail) => this with { Detail = detail, };

    public static PMSLocalIdBuilder Create(VisVersion version) => new PMSLocalIdBuilder().WithVisVersion(version);

    public bool HasCustomTag =>
        (Quantity?.IsCustom ?? false)
        || (FunctionalServices?.IsCustom ?? false)
        || (Content?.IsCustom ?? false)
        || (Position?.IsCustom ?? false)
        || (State?.IsCustom ?? false)
        || (Command?.IsCustom ?? false)
        || (MaintenanceCategory?.IsCustom ?? false)
        || (ActivityType?.IsCustom ?? false)
        || (Detail?.IsCustom ?? false);

    public bool IsValid =>
        VisVersion is not null
        && Items.PrimaryItem is not null
        && ActivityType is not null
        && (
            Quantity is not null
            || FunctionalServices is not null
            || Content is not null
            || Position is not null
            || State is not null
            || Command is not null
            || MaintenanceCategory is not null
            || ActivityType is not null
            || Detail is not null
        );

    public bool IsEmpty =>
        Items.PrimaryItem is null
        && Items.SecondaryItem is null
        && (
            Quantity is null
            && FunctionalServices is null
            && Content is null
            && Position is null
            && State is null
            && Command is null
            && MaintenanceCategory is null
            && ActivityType is null
            && Detail is null
        );

    public bool IsEmptyMetadata =>
        Quantity is null
        && FunctionalServices is null
        && Content is null
        && Position is null
        && State is null
        && Command is null
        && MaintenanceCategory is null
        && ActivityType is null
        && Detail is null;

    public IReadOnlyList<MetadataTag> MetadataTags =>
        new List<MetadataTag?>()
        {
            Quantity,
            Content,
            Position,
            State,
            Command,
            FunctionalServices,
            MaintenanceCategory,
            ActivityType,
            Detail
        }
            .Where(m => m is not null)
            .Cast<MetadataTag>()
            .ToArray();

    public bool Equals(PMSLocalIdBuilder? other)
    {
        if (other is null)
            return false;

        if (VisVersion != other.VisVersion)
            throw new InvalidOperationException("Cant compare local IDs from different VIS versions");

        return PrimaryItem == other.PrimaryItem
            && SecondaryItem == other.SecondaryItem
            && Quantity == other.Quantity
            && FunctionalServices == other.FunctionalServices
            && Content == other.Content
            && Position == other.Position
            && State == other.State
            && Command == other.Command
            && MaintenanceCategory == other.MaintenanceCategory
            && ActivityType == other.ActivityType
            && Detail == other.Detail;
    }

    public sealed override int GetHashCode()
    {
        var hashCode = new HashCode();
        hashCode.Add(PrimaryItem);
        hashCode.Add(SecondaryItem);
        hashCode.Add(Quantity);
        hashCode.Add(FunctionalServices);
        hashCode.Add(Content);
        hashCode.Add(Position);
        hashCode.Add(State);
        hashCode.Add(Command);
        hashCode.Add(MaintenanceCategory);
        hashCode.Add(ActivityType);
        hashCode.Add(Detail);
        return hashCode.ToHashCode();
    }

    public void ToString(StringBuilder builder)
    {
        if (VisVersion is null)
            throw new InvalidOperationException("No VisVersion configured on LocalId");

        string namingRule = $"/{NamingRule}/";

        builder.Append(namingRule);

        builder.Append("vis-");
        VisVersion.Value.ToVersionString(builder);
        builder.Append('/');

        Items.Append(builder, VerboseMode);

        builder.Append("meta/");

        // NOTE: order of metadatatags matter,
        // should not be changed unless changed in the naming rule/standard
        builder.AppendMeta(Quantity);
        builder.AppendMeta(Content);
        builder.AppendMeta(State);
        builder.AppendMeta(Command);
        builder.AppendMeta(FunctionalServices);
        builder.AppendMeta(MaintenanceCategory);
        builder.AppendMeta(ActivityType);
        builder.AppendMeta(Position);
        builder.AppendMeta(Detail);

        if (builder[builder.Length - 1] == '/')
            builder.Remove(builder.Length - 1, 1);
    }

    public override string ToString()
    {
        using var lease = StringBuilderPool.Get();
        var builder = lease.Builder;

        ToString(builder);

        return lease.ToString();
    }

    public PMSLocalId Build()
    {
        if (IsEmpty)
            throw new InvalidOperationException("Cant build to PMSLocalId from empty PMSLocalIdBuilder");
        if (!IsValid)
            throw new InvalidOperationException("Cant build to PMSLocalId from invalid PMSLocalIdBuilder");

        return new PMSLocalId(this);
    }
}

File: js/lib/experimental/PMSLocalId.Builder.ts
import {
    CodebookName,
    Codebooks,
    Gmod,
    GmodPath,
    LocalIdParsingErrorBuilder,
    MetadataTag,
    ParsingState,
    VisVersion,
} from "..";
import { ILocalIdBuilderGeneric } from "../ILocalIdBuilder";
import { LocalIdItems } from "../LocalId.Items";
import { Locations } from "../Location";
import { PMSLocalId } from "./PMSLocalId";
import { PMSLocalIdParser } from "./PMSLocalId.Parser";

export class PMSLocalIdBuilder
    implements ILocalIdBuilderGeneric<PMSLocalIdBuilder, PMSLocalId>
{
    public static readonly namingRule = "dnv-v2-experimental";
    public static readonly usedCodebooks = [
        CodebookName.Quantity,
        CodebookName.Content,
        CodebookName.State,
        CodebookName.Command,
        CodebookName.FunctionalServices,
        CodebookName.MaintenanceCategory,
        CodebookName.ActivityType,
        CodebookName.Position,
        CodebookName.Detail,
    ];

    public visVersion?: VisVersion;
    public verboseMode = false;
    private _items: LocalIdItems;

    public quantity?: MetadataTag;
    public content?: MetadataTag;
    public state?: MetadataTag;
    public command?: MetadataTag;
    public functionalServices?: MetadataTag;
    public maintenanceCategory?: MetadataTag;
    public activityType?: MetadataTag;
    public position?: MetadataTag;
    public detail?: MetadataTag;

    protected constructor() {
        this._items = new LocalIdItems();
    }
    public build(): PMSLocalId {
        if (this.isEmpty)
            throw new Error(
                "Cant build to PMSLocalId from empty PMSLocalIdBuilder"
            );
        if (!this.isValid){
            throw new Error(
                "Cant build to PMSLocalId from invalid PMSLocalIdBuilder"
            );
        }

        return new PMSLocalId(this);
    }

    public static create(visVersion: VisVersion): PMSLocalIdBuilder {
        return new PMSLocalIdBuilder().withVisVersion(visVersion);
    }

    public clone() {
        return Object.assign(new PMSLocalIdBuilder(), this);
    }

    public get primaryItem() {
        return this._items?.primaryItem;
    }

    public get secondaryItem() {
        return this._items?.secondaryItem;
    }

    public get isValid(): boolean {
        return (
            !!this._items.primaryItem &&
            !!this.visVersion &&
            !!this.activityType &&
            !this.isEmptyMetadata
        );
    }

    public get hasCustomTag(): boolean {
        return this.metadataTags.some((t) => t.isCustom);
    }

    public get isEmpty(): boolean {
        return (
            !this._items.primaryItem &&
            !this._items.secondaryItem &&
            this.isEmptyMetadata
        );
    }

    public get isEmptyMetadata(): boolean {
        return this.metadataTags.length === 0;
    }

    public get metadataTags(): MetadataTag[] {
        const metaDataList = [
            this.quantity,
            this.content,
            this.functionalServices,
            this.state,
            this.command,
            this.maintenanceCategory,
            this.activityType,
            this.position,
            this.detail,
        ];
        return metaDataList.filter((m) => m) as MetadataTag[];
    }

    public validate(
        errorBuilder = new LocalIdParsingErrorBuilder()
    ): LocalIdParsingErrorBuilder {
        // Add validation for primaryItem
        if (!this.primaryItem) {
            errorBuilder.push(ParsingState.PrimaryItem);
        }
        if (!this.activityType) {
            errorBuilder.push({
                type: ParsingState.MetaActivityType,
                message: "Missing activity type",
            });
        }

        return errorBuilder;
    }

    public equals(other?: PMSLocalIdBuilder, skipOther = false): boolean {
        if (!other) return false;
        if (this.visVersion !== other.visVersion)
            throw new Error(
                "Cant compare local IDs from different VisVersions"
            );

        return !!(
            (this.primaryItem?.equals(other.primaryItem) ?? true) &&
            (this.secondaryItem?.equals(other.secondaryItem) ?? true) &&
            (this.quantity?.equals(other.quantity) ?? true) &&
            (this.functionalServices?.equals(other.functionalServices) ??
                true) &&
            (this.content?.equals(other.content) ?? true) &&
            (this.position?.equals(other.position) ?? true) &&
            (this.state?.equals(other.state) ?? true) &&
            (this.command?.equals(other.command) ?? true) &&
            (this.maintenanceCategory?.equals(other.maintenanceCategory) ??
                true) &&
            (this.activityType?.equals(other.activityType) ?? true) &&
            (this.detail?.equals(other.detail) ?? true) &&
            (skipOther ? true : other.equals(this, true))
        );
    }

    public toString(builder: string[] = []): string {
        if (!this.visVersion)
            throw new Error("No VisVersion configured on LocalId");
        const namingRule = `/${PMSLocalIdBuilder.namingRule}/`;

        builder.push(namingRule);

        builder.push("vis-");
        builder.push(this.visVersion);
        builder.push("/");

        this._items.append(builder, this.verboseMode);

        builder.push("meta/");
        this.quantity?.append(builder);
        this.content?.append(builder);
        this.functionalServices?.append(builder);
        this.state?.append(builder);
        this.command?.append(builder);
        this.maintenanceCategory?.append(builder);
        this.activityType?.append(builder);
        this.position?.append(builder);
        this.detail?.append(builder);

        if (builder[builder.length - 1].endsWith("/")) {
            builder[builder.length - 1] = builder[builder.length - 1].slice(
                0,
                -1
            );
        }

        return builder.join("");
    }

    public getMetadataTag(name: CodebookName): MetadataTag | undefined {
        switch (name) {
            case CodebookName.Position:
                return this.position;
            case CodebookName.FunctionalServices:
                return this.functionalServices;
            case CodebookName.Quantity:
                return this.quantity;
            case CodebookName.State:
                return this.state;
            case CodebookName.Content:
                return this.content;
            case CodebookName.Command:
                return this.command;
            case CodebookName.MaintenanceCategory:
                return this.maintenanceCategory;
            case CodebookName.ActivityType:
                return this.activityType;
            case CodebookName.Detail:
                return this.detail;
            default:
                throw new Error(
                    `Unknown or invalid metadata tag name: ${name}`
                );
        }
    }

    public withVisVersion(visVersion: VisVersion): PMSLocalIdBuilder {
        return this.with((s) => (s.visVersion = visVersion));
    }

    public withoutVisVersion(): PMSLocalIdBuilder {
        return this.with((s) => (s.visVersion = undefined));
    }

    public tryWithVisVersion(visVersion: VisVersion): PMSLocalIdBuilder {
        if (!visVersion) return this;
        return this.withVisVersion(visVersion);
    }

    public withVerboseMode(verboseMode: boolean): PMSLocalIdBuilder {
        return this.with((s) => (s.verboseMode = verboseMode));
    }

    public withoutVerboseMode(): PMSLocalIdBuilder {
        return this.with((s) => (s.verboseMode = false));
    }

    public withPrimaryItem(item: GmodPath): PMSLocalIdBuilder {
        return this.with((s) => (s._items.primaryItem = item));
    }

    public withoutPrimaryItem(): PMSLocalIdBuilder {
        return this.with((s) => (s._items.primaryItem = undefined));
    }

    public tryWithPrimaryItem(item?: GmodPath): PMSLocalIdBuilder {
        if (!item) return this;
        return this.withPrimaryItem(item);
    }

    public withSecondaryItem(item: GmodPath): PMSLocalIdBuilder {
        return this.with((s) => (s._items.secondaryItem = item));
    }

    public withoutSecondaryItem(): PMSLocalIdBuilder {
        return this.with((s) => (s._items.secondaryItem = undefined));
    }

    public tryWithSecondaryItem(item?: GmodPath): PMSLocalIdBuilder {
        if (!item) return this;
        return this.withSecondaryItem(item);
    }

    public withQuantity(metadataTag: MetadataTag): PMSLocalIdBuilder {
        return this.with((s) => (s.quantity = metadataTag));
    }

    public withFunctionalServices(metadataTag: MetadataTag): PMSLocalIdBuilder {
        return this.with((s) => (s.functionalServices = metadataTag));
    }

    public withContent(metadataTag: MetadataTag): PMSLocalIdBuilder {
        return this.with((s) => (s.content = metadataTag));
    }

    public withPosition(metadataTag: MetadataTag): PMSLocalIdBuilder {
        return this.with((s) => (s.position = metadataTag));
    }

    public withState(metadataTag: MetadataTag): PMSLocalIdBuilder {
        return this.with((s) => (s.state = metadataTag));
    }

    public withCommand(metadataTag: MetadataTag): PMSLocalIdBuilder {
        return this.with((s) => (s.command = metadataTag));
    }

    public withMaintenanceCategory(
        metadataTag: MetadataTag
    ): PMSLocalIdBuilder {
        return this.with((s) => (s.maintenanceCategory = metadataTag));
    }

    public withActivityType(metadataTag: MetadataTag): PMSLocalIdBuilder {
        return this.with((s) => (s.activityType = metadataTag));
    }

    public withDetail(metadataTag: MetadataTag): PMSLocalIdBuilder {
        return this.with((s) => (s.detail = metadataTag));
    }

    public tryWithMetadataTag(metadataTag?: MetadataTag): PMSLocalIdBuilder {
        if (!metadataTag) return this;
        return this.withMetadataTag(metadataTag);
    }

    public withMetadataTag(metadataTag: MetadataTag): PMSLocalIdBuilder {
        switch (metadataTag.name) {
            case CodebookName.Position:
                return this.withPosition(metadataTag);
            case CodebookName.FunctionalServices:
                return this.withFunctionalServices(metadataTag);
            case CodebookName.Quantity:
                return this.withQuantity(metadataTag);
            case CodebookName.State:
                return this.withState(metadataTag);
            case CodebookName.Content:
                return this.withContent(metadataTag);
            case CodebookName.Command:
                return this.withCommand(metadataTag);
            case CodebookName.MaintenanceCategory:
                return this.withMaintenanceCategory(metadataTag);
            case CodebookName.ActivityType:
                return this.withActivityType(metadataTag);
            case CodebookName.Detail:
                return this.withDetail(metadataTag);
            default:
                throw new Error(
                    `Unknown or invalid metadata tag name: ${metadataTag.name}`
                );
        }
    }

    public withoutMetadataTag(name: CodebookName): PMSLocalIdBuilder {
        switch (name) {
            case CodebookName.Position:
                return this.withoutPosition();
            case CodebookName.FunctionalServices:
                return this.withoutFunctionalServices();
            case CodebookName.Quantity:
                return this.withoutQuantity();
            case CodebookName.State:
                return this.withoutState();
            case CodebookName.Content:
                return this.withoutContent();
            case CodebookName.Command:
                return this.withoutCommand();
            case CodebookName.MaintenanceCategory:
                return this.withoutMaintenanceCategory();
            case CodebookName.ActivityType:
                return this.withoutActivityType();
            case CodebookName.Detail:
                return this.withoutDetail();
            default:
                throw new Error(
                    `Unknown or invalid metadata tag name: ${name}`
                );
        }
    }

    private withoutPosition() {
        return this.with((s) => (s.position = undefined));
    }

    private withoutQuantity() {
        return this.with((s) => (s.quantity = undefined));
    }
    private withoutContent() {
        return this.with((s) => (s.content = undefined));
    }
    private withoutCommand() {
        return this.with((s) => (s.command = undefined));
    }
    private withoutState() {
        return this.with((s) => (s.state = undefined));
    }
    private withoutFunctionalServices() {
        return this.with((s) => (s.functionalServices = undefined));
    }
    private withoutMaintenanceCategory() {
        return this.with((s) => (s.maintenanceCategory = undefined));
    }
    private withoutActivityType() {
        return this.with((s) => (s.activityType = undefined));
    }
    private withoutDetail() {
        return this.with((s) => (s.detail = undefined));
    }

    public with(u: { (state: PMSLocalIdBuilder): void }): PMSLocalIdBuilder {
        const n = this.clone();
        u && u(n);
        return n;
    }

    public static parse(
        localIdStr: string | undefined,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return PMSLocalIdParser.parse(
            localIdStr,
            gmod,
            codebooks,
            locations,
            errorBuilder
        );
    }

    public static async parseAsync(
        localIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return PMSLocalIdParser.parseAsync(localIdString, errorBuilder);
    }

    public static tryParse(
        localIdStr: string | undefined,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return PMSLocalIdParser.tryParse(
            localIdStr,
            gmod,
            codebooks,
            locations,
            errorBuilder
        );
    }

    public static async tryParseAsync(
        localIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return PMSLocalIdParser.tryParseAsync(localIdString, errorBuilder);
    }
}
